From d08e9d8383ab5c3d36fefc031b7a4be39c338657 Mon Sep 17 00:00:00 2001
From: Don Darling <don.osc2@gmail.com>
Date: Fri, 30 Apr 2010 12:56:58 -0500
Subject: [PATCH 09/14] v4l2src: try progressive mode first for component input on some platforms

Normally, v4l2src tries INTERLACED mode first on a capture input, and if that
fails, it tries progressive.  However, on DM6467 the component capture driver
will accept INTERLACED, but subsequent calls to dqbuf will fail.  On DM6467
component input, we must try NONE (Progressive) on the first try.
---
 sys/v4l2/gstv4l2object.c |   48 ++++++++++++++++++++++++++++++++++++++++++++++
 1 files changed, 48 insertions(+), 0 deletions(-)

diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index ac0b0d7..e02dc1d 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -1790,6 +1790,30 @@ gst_v4l2_object_get_nearest_size (GstV4l2Object * v4l2object,
   fmt.fmt.pix.pixelformat = pixelformat;
   fmt.fmt.pix.field = V4L2_FIELD_INTERLACED;
 
+  /* Normally, v4l2src tries INTERLACED mode first on a capture input, and if
+   * that fails, it tries progressive.  However, on DM6467 the component
+   * capture driver will accept INTERLACED, but subsequent calls to dqbuf will
+   * fail.  On DM6467 component input, we must try NONE (Progressive) on the
+   * first try.
+   */
+  {
+    int component_input_id;
+    int current_input_id;
+
+    #if defined(Platform_dm6467) || defined(Platform_dm6467t)
+      component_input_id = 1;
+    #else
+      component_input_id = -1;
+    #endif
+
+    if (component_input_id > 0) {
+      gst_v4l2_get_input(v4l2object, &current_input_id);
+      if (current_input_id == component_input_id) {
+          fmt.fmt.pix.field = V4L2_FIELD_NONE;
+      }
+    }
+  }
+
   r = v4l2_ioctl (fd, VIDIOC_TRY_FMT, &fmt);
   if (r < 0 && (errno == EINVAL || errno == EPERM)) {
     /* try again with progressive video */
@@ -1881,6 +1905,30 @@ gst_v4l2_object_set_format (GstV4l2Object * v4l2object, guint32 pixelformat,
    *  combined, there are other modes for requesting fields individually) */
   format.fmt.pix.field = V4L2_FIELD_INTERLACED;
 
+  /* Normally, v4l2src tries INTERLACED mode first on a capture input, and if
+   * that fails, it tries progressive.  However, on DM6467 the component
+   * capture driver will accept INTERLACED, but subsequent calls to dqbuf will
+   * fail.  On DM6467 component input, we must try NONE (Progressive) on the
+   * first try.
+   */
+  {
+    int component_input_id;
+    int current_input_id;
+
+    #if defined(Platform_dm6467) || defined(Platform_dm6467t)
+      component_input_id = 1;
+    #else
+      component_input_id = -1;
+    #endif
+
+    if (component_input_id > 0) {   
+      gst_v4l2_get_input(v4l2object, &current_input_id);
+      if (current_input_id == component_input_id) {
+        format.fmt.pix.field = V4L2_FIELD_NONE;
+      }
+    }
+  }
+
   #if defined(MVL_LSP) 
     v4l2object->bytesused = format.fmt.pix.sizeimage;
   #endif
-- 
1.6.3.3

